test/other/embind_tsgen_es6.d.ts

```diff
@@ -0,0 +1,113 @@
+  constFn(): number;
+  longFn(_0: number): number;
+  functionThree(_0: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): number;
+  functionSix(str: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): number;
+  delete(): void;
+}
+
+export interface BarValue<T extends number> {
+  value: T;
+}
+export type Bar = BarValue<0>|BarValue<1>|BarValue<2>;
+
+export interface EmptyEnumValue<T extends number> {
+  value: T;
+}
+export type EmptyEnum = never/* Empty Enumerator */;
+
+export type ValArrIx = [ Bar, Bar, Bar, Bar ];
+
+export interface IntVec {
+  push_back(_0: number): void;
+  resize(_0: number, _1: number): void;
+  size(): number;
+  get(_0: number): number | undefined;
+  set(_0: number, _1: number): boolean;
+  delete(): void;
+}
+
+export interface MapIntInt {
+  keys(): IntVec;
+  get(_0: number): number | undefined;
+  set(_0: number, _1: number): void;
+  size(): number;
+  delete(): void;
+}
+
+export interface Foo {
+  process(_0: Test): void;
+  delete(): void;
+}
+
+export type ValObj = {
+  foo: Foo,
+  bar: Bar
+};
+
+export interface ClassWithConstructor {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface ClassWithTwoConstructors {
+  delete(): void;
+}
+
+export interface ClassWithSmartPtrConstructor {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface BaseClass {
+  fn(_0: number): number;
+  delete(): void;
+}
+
+export interface DerivedClass extends BaseClass {
+  fn2(_0: number): number;
+  delete(): void;
+}
+
+export type ValArr = [ number, number, number ];
+
+interface EmbindModule {
+  Test: {staticFunction(_0: number): number; staticFunctionWithParam(x: number): number; staticProperty: number};
+  class_returning_fn(): Test;
+  class_unique_ptr_returning_fn(): Test;
+  a_class_instance: Test;
+  an_enum: Bar;
+  Bar: {valueOne: BarValue<0>, valueTwo: BarValue<1>, valueThree: BarValue<2>};
+  EmptyEnum: {};
+  enum_returning_fn(): Bar;
+  IntVec: {new(): IntVec};
+  MapIntInt: {new(): MapIntInt};
+  Foo: {};
+  ClassWithConstructor: {new(_0: number, _1: ValArr): ClassWithConstructor};
+  ClassWithTwoConstructors: {new(): ClassWithTwoConstructors; new(_0: number): ClassWithTwoConstructors};
+  ClassWithSmartPtrConstructor: {new(_0: number, _1: ValArr): ClassWithSmartPtrConstructor};
+  BaseClass: {};
+  DerivedClass: {};
+  a_bool: boolean;
+  an_int: number;
+  optional_test(_0: Foo | undefined): number | undefined;
+  global_fn(_0: number, _1: number): number;
+  smart_ptr_function(_0: ClassWithSmartPtrConstructor): number;
+  smart_ptr_function_with_params(foo: ClassWithSmartPtrConstructor): number;
+  function_with_callback_param(_0: (message: string) => void): number;
+  string_test(_0: ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string): string;
+  wstring_test(_0: string): string;
+}
+export type MainModule = WasmModule & EmbindModule;
+export default function MainModuleFactory (options?: unknown): Promise<MainModule>;
```

Comment:
This file should end in a newline.

Reply:
This can all happen in followup PRs, but I wonder how `EXPORT_NAME` should interact with this.

Should we invert this such that we have:

```
export type MainModuleType = WasmModule & EmbindModule;
export default function MainModule (options?: unknown): Promise<MainModuleType>;
```

Also, I'm not sure what `MainModule` is here.. we don't use the term `MainModule` anywhere else in emscripten so it seems a little odd.  @brendandahl are you planning on changing this?

Reply:
This also seemed iffy to me, as right now my consuming code has to do `import { type MainModule as MyModule } from "my-package";` to look as expected (and to allow me to import multiple wasm modules in the same file.

When this PR is merged I can take a stab at using `EXPORT_NAME` instead of a hard-coded default

Reply:
Re: your question about inverting it. To me at least, it's clearer for the type to be named `MainModule`, and the default factory-function export be named as such. But I don't have a strong opinion on that 🤷  -- my only preference would be to align with some TypeScript standard naming convention for this kind of situation, if such a thing exists.

Reply:
The actual name of the _default_ export is irrelevant, as with all default exports (as far as I know). For this reason, I think `EXPORT_NAME` doesn't currently do anything meaningful with `EXPORT_ES6` and `MODULARIZE` - it probably could [rename the type for you](https://github.com/emscripten-core/emscripten/issues/20952), but I think that's a separate thing as noted. It's also a little awkward, because in some situations the EXPORT_NAME is the name of the _factory_, (e.g. `createMyModule`) as opposed to the name of the thing it returns (e.g. `MyModule`), so it almost deserves to be a separate option...

I probably would suggest leaving the type exported as `MainModule` (or maybe later, some configurable thing). It's not typical to have types end in a `Type` suffix, anymore than you'd end a class with a `Class` suffix.

Reply:
I agree its a separate issue.  But why not `MainModule` -> EXPORT_NAME here.

That would make the default `Module` rather than `MainModule` which fits much better with existing emscripten concept of `Module` as something that is just the default value of `EXPORT_NAME`.

Certainly if I build a module with `EXPORT_NAME=Foo` I would expect something `Foo` to replace `Module` in most places as the public facing name.

Reply:
It's been a bit since I tried different combinations of these options, and I remember it being a bit confusing to me how `EXPORT_NAME`, `MODULARIZE`, and `EXPORT_ES6` all interplay with each other... but I think in some situations, `EXPORT_NAME` ends up being the name of the _factory function_ that you call to instantiate the module (e.g. the docs suggest doing `EXPORT_NAME=createMyModule` so that you can do `const module = await createMyModule();`). So now, overloading that same option to rename the type of the thing that is returned from that factory would be pretty strange... at the very least it would be tricky to try to explain what this option does in context of other options. Even if you appended `Type` to it, now in the example from the docs, you'd end up with `createMyModuleType` which is probably not what you expect.

Reply:
Wouldn't the logical thing to do be to name the factor function `MainModuleFactory` as `EXPORT_NAME` since that would match the existing behaviour e.g. `createMyModule(..)`  rather than `MainModuleFactory`?

Reply:
> Wouldn't the logical thing to do be to name the factory function MainModuleFactory as EXPORT_NAME

It makes some sense from this end, but since it's the default export, its name makes 0 difference to the consumer. You'll end up doing `import WHATEVER_NAME_I_WANT from 'my-module'`, and the `EXPORT_NAME` doesn't actually do anything from the consumer's end, since default export names don't matter. In fact, you can drop the name `MainModuleFactory` all together and just do `export default function (options?: unknown): Promise<MainModule>;` to make that more explicit.

Here's an analogy (to the best of my understanding) of the situation we end up in if you use `EXPORT_NAME`, `MODULARIZE`, and `EXPORT_ES6` - note that `EXPORT_NAME` actually ends up being irrelevant to the consumer (though I think it is relevant if you don't use `EXPORT_ES6` - that's a different situation):
![image](https://github.com/emscripten-core/emscripten/assets/8780033/68e9c2c2-486b-4b91-94ab-e9cba691a617)

I see a few possibilities of how this could all ideally work

1. Make a breaking changes to what `EXPORT_NAME` does - now you give it the _name of your module itself_, so that `EXPORT_NAME=MyModule` would give you a factory called `createMyModule` (or `create_wasm_module_MyModule` or some other name to try to get uniqueness) and we re-use `EXPORT_NAME` to name the type as well. Note that if you don't use `EXPORT_ES6`, the name of this factory _does_ matter, so we have to consider both cases.
2. Accept that `EXPORT_NAME` does totally different things depending on whether you use `EXPORT_ES6` or not - in one case, it renames the factory, and in the other case, it renames the type. This seems pretty confusing to me.
3. Introduce a whole new setting like `MODULE_NAME` so that we can actually [rename the module type name](https://github.com/emscripten-core/emscripten/issues/20952)

Or I guess:
4. Just accept that you can't rename `MainModule` - it's not that annoying to just alias it on the consuming side.

Reply:
Filed #21470 in an attempt to capture this conversation.

Reply:
Thanks for the clear case.   I agree we don't need to honor `EXPORT_NAME` here.

I wonder if we can/should just make `EXPORT_NAME` into and an error when used with `EXPORT_ES6` if it really doesn't do anything in this mode.